### DISCUSSION TITLE
MTU size for Nordic Uart service

### DISCUSSION THREAD

**Discussion by linuslundin** [ThomasGerstenberg/blatann#91] [Q&A] [answered]
When I use the Nordic UART service the packets sent will only be MTU_SIZE_DEFAULT = 23 bytes.
I have previously negotiated a much larger MTU size but the uart service seems to ignore this.

    peer = ble_device.connect(target_address).wait()
    if not peer:
        logging.warning("Timed out connecting to device")
        quit()
    logging.info("Connected, conn_handle: {}".format(peer.conn_handle))
    logging.info("Exchanging MTU")
    peer.exchange_mtu(peer.max_mtu_size).wait(10)
    logging.info("MTU Exchange complete, mtu_size:{} discovering services".format(peer.mtu_size))
    
    logging.info("1 Current MTU size = %d",peer.max_mtu_size)
    ### Above log prints: INFO:root:1 Current MTU size = 247

    # Initiate service discovery and wait for it to complete
    _, event_args = peer.discover_services().wait(exception_on_timeout=False)
    logging.info("Service discovery complete! status: {}".format(event_args.status))
    uart_service = nordic_uart.find_nordic_uart_service(peer.database)

    logging.info("2 Current MTU size = %d",uart_service.max_write_length)
    ### Above log prints: INFO:root:2 Current MTU size = 23

    if not uart_service:
        logging.info("Failed to find Nordic UART service in peripheral database")
        peer.disconnect().wait()
        ble_device.close()
        quit()

    logging.info("3 Current MTU size = %d",uart_service.max_write_length)
    ### Above log prints: INFO:root:3 Current MTU size = 23

    # Initialize the service
    uart_service.initialize().wait(5)
    uart_service.on_data_received.register(on_data_rx)

I have changed the nordic_urat/service.py init function to this:

        def __init__(self, service, max_characteristic_size=None):
        """
        :type service: blatann.gatt.gattc.GattcService
        """
        self._service = service
        #self._server_characteristic_size = MTU_SIZE_DEFAULT
        if max_characteristic_size is None:
            self._server_characteristic_size = self._service.peer.max_mtu_size - WRITE_BYTE_OVERHEAD
        else:
            self._server_characteristic_size = max_characteristic_size

        self._tx_char = self._service.find_characteristic(NORDIC_UART_TX_CHARACTERISTIC_UUID)
        self._rx_char = self._service.find_characteristic(NORDIC_UART_RX_CHARACTERISTIC_UUID)
        self._feature_char = self._service.find_characteristic(NORDIC_UART_FEATURE_CHARACTERISTIC_UUID)
        self._init_complete_event = EventSource("Init Complete", logger)
        self._on_data_received = EventSource("Data Received", logger)
        self._on_write_complete = EventSource("Write Complete", logger)

After this it works just fine, is this a good way of fixing this issue or am I missing something in how I use the driver?

Regards

**Comment by ThomasGerstenberg** [comment, marked as answer]
Sorry for the delayed response. You're correct, I wrote the UART service before I added support for larger MTUs and didn't go back to fix it up. Your changes make sense to me

**Comment by linuslundin** [comment]
Good, I can see that the main branch is a bit different then my code, will you implement a fix or should I try to create a pull request?
